Sensors/SMCDellSensors/SMIMonitor.cpp

```diff
@@ -246,77 +244,141 @@ void SMIMonitor::createShared() {
+		IOLockLock(mainLock);
+		thread_call_enter(updateCall);
+		
+		while (initialized == -1) {
+			IOLockSleep(mainLock, &initialized, THREAD_UNINT);
+		}
+		IOLockUnlock(mainLock);
+
+		if (initialized != KERN_SUCCESS) {
 			success = false;
+			break;
 		}
-		
+
 		DBGLOG("sdell", "found %u fan sensors and %u temp sensors", fanCount, tempCount);
+		break;
+	}
 
-		if (!success) {
-			OSSafeReleaseNULL(workloop);
-			OSSafeReleaseNULL(timerEventSource);
+	if (!success) {
+		if (updateCall) {
+			while (!thread_call_free(updateCall))
+				thread_call_cancel(updateCall);
+			updateCall = nullptr;
 		}
 	}
 	
 	DBGLOG("sdell", "Based on I8kfan project and adopted to VirtualSMC plugin");
 
-	IOLockUnlock(mainLock);
 	return success;
 }
 
 void SMIMonitor::start() {
-	IOLockLock(mainLock);
-	if (!initialUpdateSensors) {
-		for (int i=0; i<fanCount; ++i)
-			i8k_set_fan_control_auto(state.fanInfo[i].index); // force automatic control
-		updateSensors();
-		initialUpdateSensors = true;
-	}
-	IOLockUnlock(mainLock);
 }
 
 void SMIMonitor::handlePowerOff() {
-	timerEventSource->disable();
 }
 
 void SMIMonitor::handlePowerOn() {
-	timerEventSource->enable();
+}
+
+bool SMIMonitor::postSmcUpdate(SMC_KEY key, size_t index, const void *data, uint32_t dataSize)
+{
+	IOSimpleLockLock(queueLock);
+
+	bool success = false;
+	while (1) {
+	
+		if (dataSize > sizeof(StoredSmcUpdate::data)) {
+			SYSLOG("sdell", "postRequest dataSize overflow %u", dataSize);
+			break;
+		}
+
+		for (size_t i = 0; i < storedSmcUpdates.size() && !success; i++) {
+			auto &si = storedSmcUpdates[i];
+			if (si.key == key && si.index == index) {
+				si.size = dataSize;
+				if (dataSize > 0)
+					lilu_os_memcpy(si.data, data, dataSize);
+				success = true;
+				break;
+			}
+		}
+		
+		if (success) break;
+
+		if (storedSmcUpdates.size() == MaxActiveSmcUpdates) {
+			SYSLOG("sdell", "postRequest reserve overflow");
+			break;
+		}
+
+		StoredSmcUpdate si {key, index, dataSize};
+		if (dataSize > 0)
+			lilu_os_memcpy(si.data, data, dataSize);
+		if (storedSmcUpdates.push_back(si)) {
+			success = true;
+			break;
+		}
+		break;
+	}
+	IOSimpleLockUnlock(queueLock);
+	
+	if (!success)
+		SYSLOG("sdell", "unable to store smc update");
+
+	return success;
+}
+
+IOReturn SMIMonitor::bindCurrentThreadToCpu0()
+{
+	// Obtain power management callbacks
+	pmCallBacks_t callbacks {};
+	pmKextRegister(PM_DISPATCH_VERSION, nullptr, &callbacks);
+	
+	if (!callbacks.LCPUtoProcessor) {
+		SYSLOG("sdell", "failed to obtain LCPUtoProcessor");
+		return KERN_FAILURE;
+	}
+	
+	if (!callbacks.ThreadBind) {
+		SYSLOG("sdell", "failed to obtain ThreadBind");
+		return KERN_FAILURE;
+	}
+
+	auto enable = ml_set_interrupts_enabled(FALSE);
+	
+	auto processor = callbacks.LCPUtoProcessor(0);
+	if (processor == nullptr) {
+		ml_set_interrupts_enabled(enable);
+		SYSLOG("sdell", "failed to call LCPUtoProcessor with cpu 0");
+		return KERN_FAILURE;
+	}
+	callbacks.ThreadBind(processor);
```

Comment:
It's insufficient to only disable interrupts when calling this. Preemption also needs to be disabled when calling ThreadBind, otherwise a wrong thread may be bound, see https://github.com/07151129/ThreadBinder/commit/ab3f4bad83866d97d51bcd8502b6e9d611ea5d6c#diff-8f63a794a20b44157c6433d7df606316

Reply:
@07151129, thank you for noticing, it has been fixed.

Reply:
Sorry, I was wrong, VirtualSMCAPI::encodeFp and VirtualSMCAPI::decodeFp work as expected.